### PR TITLE
Refactor: Simplify publishes calls

### PR DESCRIPTION
**Pull request by Rgascoin**
Refactor: Users have to provide optional field that can be simplified.
Solution: simplify publishes methods with default values

\+ Try to improve syntax on publishing methods.

ref: https://github.com/aleph-im/aleph-sdk-ts/issues/62

### CODE REVIEW

File: src/utils/messageBuilder.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import { ItemType } from "../messages/message";
+import { Account } from "../accounts/account";
+
+type MessageBuilderConfig<T, C> = {
```
Comment (BjrInt) [inline src/utils/messageBuilder.ts:4]: Using `T` for `content` and `C` for `type` might be a bit misleading, the contrary would make much more sense

File: src/messages/create/publish.ts [hunk 2]
```diff
@@ -46,18 +46,13 @@ type PushFileConfiguration = {
...
-            configuration.message.item_content = serialized;
-            configuration.message.item_hash = new shajs.sha256().update(serialized).digest("hex");
-        }
-    }
-    if (!configuration.inlineRequested) {
+    if (serialized.length < 50000 && configuration.inlineRequested) {
+        configuration.message.item_type = ItemType.inline;
+        configuration.message.item_content = serialized;
+        configuration.message.item_hash = new shajs.sha256().update(serialized).digest("hex");
+    } else {
```
Comment (BjrInt) [inline src/messages/create/publish.ts:55]: Would make sense to either `throw` an error or issue a warning if the message payload is longer than 50k characters (or bytes?).

Also what happens if I chose this, with a payload heavier than the limit? 
```js
{
  storage: 'inline',
  inlineRequested: true
}
```

the condition below uses the "storageEngine" chosen by the user.

File: src/utils/messageBuilder.ts (hunk 1 above)
Comment (Rgascoin) [reply src/utils/messageBuilder.ts]: Updated in [ea6951c](https://github.com/aleph-im/aleph-sdk-ts/pull/110/commits/ea6951cb62e03149743e73afbea87cf4d58e27de)

File: tests/messages/forget/publish.test.ts [hunk 3]
```diff
@@ -23,11 +23,9 @@ describe("Forget publish tests", () => {
         });
 
         const Fres = await forget.Publish({
-            APIServer: DEFAULT_API_V2,
             channel: "TEST",
             hashes: [res.item_hash],
             inlineRequested: true,
```
[the remark this answers is not in the document]
Comment (Rgascoin) [reply tests/messages/forget/publish.test.ts:28]: According to the documentation Forget message can be either inline or storage/ipfs

I added a test with another storage engine for the demo: [46f8b32](https://github.com/aleph-im/aleph-sdk-ts/pull/110/commits/46f8b32003e3a554c7e2df05aa127cb0fba4517d)

**Comment by Rgascoin** [comment]
## Summarizing this PR:
- Add default values to simplify publishes methods (Default_API, storage location).
- Deprecated `inlineRequested` over `storageEngine` param.
- clean-up some code

## Recap of the storage location of item_content:
Aggregate:
- Message content can be fill as: Inline/Storage/IPFS
- If message content size > MAX_SIZE and storage is inline:
	- message content is stored on Storage

Post:
- Message content can be fill as: Inline/Storage/IPFS
- If message content size > MAX_SIZE and storage is inline:
	- message content is stored on Storage

Forget:
- Message content can be fill as: Inline/Storage/IPFS
- If message content size > MAX_SIZE and storage is inline:
	- message content is stored on Storage

Store:
- Uploaded file can be fill as: Storage/IPFS
- Message content is store on: Inline
- If message content size > MAX_SIZE:
	- message content is stored on Storage

Program:
- Uploaded code is store on: Storage
- Message content is store on: inline
- If message content size > MAX_SIZE:
	- message content is stored on Storage or IPFS

File: src/messages/aggregate/publish.ts [hunk 4]
```diff
@@ -42,41 +46,45 @@ type AggregatePublishConfiguration<T> = {
...
  * @param configuration The configuration used to publish the aggregate message.
  */
-export async function Publish<T>(configuration: AggregatePublishConfiguration<T>): Promise<AggregateMessage<T>> {
+export async function Publish<T>({
+    account,
+    address,
+    key,
+    content,
+    channel,
+    storageEngine = ItemType.inline,
```
Comment (hoh) [inline src/messages/aggregate/publish.ts:55]: Can the default not be configured in the AggregatePublishConfiguration type in TS ?

File: src/messages/forget/publish.ts [hunk 5]
```diff
@@ -38,42 +40,44 @@ type ForgetPublishConfiguration = {
...
     const timestamp = Date.now() / 1000;
-    const content: ForgetContent = {
-        address: configuration.account.address,
+    const forgetContent: ForgetContent = {
+        address: account.address,
         time: timestamp,
-        hashes: configuration.hashes,
-        reason: configuration.reason ? configuration.reason : "None",
+        hashes: hashes,
+        reason: reason || "None",
```
Comment (hoh) [inline src/messages/forget/publish.ts:59]: Why "None" as a string ?

**Comment by hoh** [comment]
Program: Uploaded code is store on: Ideally IPFS by default.

Aggregate: Message content can be fill as: Inline/Storage/IPFS, with a preference/default for inline.

**Comment by hoh** [comment]
Program execution does not benefit as much from the built-in Storage option (still needs to download the data after fetching the message), so IPFS should be encouraged.

File: src/messages/forget/publish.ts (hunk 5 above)
Comment (Rgascoin) [reply src/messages/forget/publish.ts:59]: Not sure I remember the original reason for this behavior.
Should this field be `undefined` if there is no reason ?

**Comment by odesenfans** [comment]
Hey @Rgascoin , not sure I get some of the cases you described above. Anyway, here's my take:

* The default for all message contents should **always** to be inline as long as we're below `MAX_INLINE_SIZE` and unless the user requests it specifically. See point 1 below. Therefore, there should be a boolean parameter named something like `allow_content_inlining` which should be true by default, and most users should just never worry about this parameter.
* In the publish function, check the content size against `MAX_INLINE_SIZE` and use the Aleph storage engine if needed. I don't see a valid use case for an Aleph message content to be stored on IPFS.
* Whether to use IPFS or the Aleph storage engine for files is up to you. Using the storage engine has its advantages (ex: the file is guaranteed to be available on at least one node). IPFS is better for really big files. Follow @hoh 's advice on this.

1. The only case where this makes sense is when a user sends a message that he may want to forget later on. Inline message contents are stored as part of the message sync dumps, and therefore can never be truly forgotten.

File: src/messages/forget/publish.ts (hunk 5 above)
Comment (Rgascoin) [reply src/messages/forget/publish.ts]: This should fix this: https://github.com/aleph-im/aleph-sdk-ts/pull/110/commits/9e4f52e5e05b4c9208b013f5bba8143d81183b4c

**Comment by Rgascoin** [comment]
Thanks for your explanations @odesenfans, I updated it to allow user to choose a non-inline storage if needed, by default every message are stored inlined.
According to @hoh recommendation, Program source codes are now stored on IPFS by default

File: src/messages/forget/publish.ts (hunk 5 above)
Comment (BjrInt) [reply src/messages/forget/publish.ts]: I guess this is due to the fact that "None" is the Python null type.

**Review by BjrInt — APPROVED** [review]
LGTM

File: src/messages/forget/publish.ts (hunk 5 above)
Comment (odesenfans) [reply src/messages/forget/publish.ts]: This field should remain null if it's null. "None" ends up stored as a string.

**Pull request merged**